WcaOnRails/app/models/registration.rb

```diff
@@ -86,6 +83,10 @@ def best_solve(event, type)
     person ? person.best_solve(event, type) : SolveTime.new(event.id, type, 0)
   end
 
+  def event_picker_events
+    registration_events.map(&:event_object).sort_by(&:rank)
+  end
```

Comment:
Right, I see the problem.
At the moment I think we can at least get rid of these `event_object` methods and use `event` instead. Note that even if you have an unsaved `competition_event` with `event_id` set, then `competition_event.event` will generate appropriate query with this `event_id` and the result will be the same as `competition_event.event_object`.

I have one idea that should let us avoid these `event_picker_events` methods entirely:
```ruby
def events
  if competition_events.any?(&:new_record?)
    competition_events.map(&:event)
  else
   super
  end
end
```
The same would go for registrations. I'm not sure how good would that be in wider context. @larspetrus what do you think? Have you stumbled upon such a problem before? (unsaved records from `accepted_nested_attributes_for` and `through` association).